000X-B-C-I-R-S-error-handling-codes.md

```diff
@@ -0,0 +1,307 @@
+
+This makes troubleshooting and debugging quite difficult and requires a deep understanding of the parts of Dapr and how those parts interact with each other.
+
+To help with the issues raised above, it would be ideal if a solution could provide:
+- Greater details of errors that occured.
+- Error details in a structured format.
+- Consistency in the error details.
+- An indication where within the Dapr execution (Init, Runtime, Components, SDKs, etc...) the error occurred.
+
+## Background
+
+## Related Items
+
+### Related proposals 
+
+
+### Related issues 
+
+https://github.com/dapr/dapr/issues/6068
+
+## Expectations and alternatives
+
+## Implementation Details
+
+# Design
+
+## Solution
+Utilize and follow the [gRPC Richer Error Model](https://grpc.io/docs/guides/error/#richer-error-model) and [Google API Errors Model in the Design Guide](https://cloud.google.com/apis/design/errors#error_model)
+
+### Error Code Standard
+The [Google API Error Model](https://cloud.google.com/apis/design/errors#error_model) has the following Protobuf format:
+```go
+package google.rpc;
+
+// The `Status` type defines a logical error model that is suitable for
+// different programming environments, including REST APIs and RPC APIs.
+message Status {
+  // A simple error code that can be easily handled by the client. The
+  // actual error code is defined by `google.rpc.Code`.
+  int32 code = 1;
+
+  // A developer-facing human-readable error message in English. It should
+  // both explain the error and offer an actionable resolution to it.
+  string message = 2;
+
+  // Additional error information that the client code can use to handle
+  // the error, such as retry info or a help link.
+  repeated google.protobuf.Any details = 3;
+}
+```
+
+Here is one of the possible details that can be added to the above
+error structure. This is defined in the [error_details.proto Protobuf](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto) 
+
+```go
+message ErrorInfo {
+  // The reason of the error. This is a constant value that identifies the
+  // proximate cause of the error. Error reasons are unique within a particular
+  // domain of errors. This should be at most 63 characters and match a
+  // regular expression of `[A-Z][A-Z0-9_]+[A-Z0-9]`, which represents
+  // UPPER_SNAKE_CASE.
+  string reason = 1;
+
+  // The logical grouping to which the "reason" belongs. The error domain
+  // is typically the registered service name of the tool or product that
+  // generates the error. Example: "pubsub.googleapis.com". If the error is
+  // generated by some common infrastructure, the error domain must be a
+  // globally unique value that identifies the infrastructure. For Google API
+  // infrastructure, the error domain is "googleapis.com".
+  string domain = 2;
```

Comment:
is it going to be only `dapr.io` here? And if so do we need the domain field?

Reply:
For now we're planning it to be `dapr.io`.

Reply:
ErrorInfo is not out structure, it is Google's gRPC and we will pass `dapr.io`, so SDKs can know which ones they care about in the array of ErrorInfo.